### PR TITLE
fix: fix a panic during physical plan generation

### PR DESCRIPTION
**Pull request by SteveLauC**
## What does this PR do

Fix a panic that happens while generating physical plan from a logical plan, our assertion here is not that accurate, when `input_dfschema` does not contain an `_id` field (the `else` branch), it does mean that the query does not involve the `_id` field, so we will add an extra `_id` field to the `ParquetExec`, then `input_schema` will have one more field than `input_dfschema`, but it doesn't mean that such a case will **always be true and permanent**.

> Difference between `dfschema` and `schema`:
>
> `dfschema` stands for DataFusion schema, you can basically think it is schema that comes from a logical plan, and `schema` comes from physical plan.

Let's demo this with an example:

```sql
SELECT COUNT(*) FROM table GROUP By foo
```

The above query will create the following logical plan:

```
Projection: COUNT(*)                              (dfschema: COUNT(*))
  Aggregate: groupBy=[[t.foo]], aggr=[[COUNT(*)]] (dfschema: foo, COUNT(*))
    TableScan:                                    (dfschema: foo)
```

And it will be converted into this physical plan:

> This is the initial physical plan, not the optimized one

```
ProjectionExec                                     (schema: COUNT(*))
    AggregationExec                                (schema: foo, COUNT(*))
        AggregationExec                            (schema: foo, COUNT(*))
            ParquetExec                            (schema: foo, _id)
```

Converting a logical plan to a physical plan is a recursive process, if we demo it with pseudocode, it will be something like this:

```
Projection_to_ProjectionExec() {
    Aggregate_to_AggregateExec() {
        TableScan_to_ParquetExec() {
            // xxxx
        }
    }    
}
```

We can see that `ParquetExec` has one more field than `TableScan`, so when converting `Aggregate` to an `AggregateExec`,

> `input_schema` means its input's schema, a node's input node is its next node, e.g., `Project`'s input is `Aggregate`

their `input_schema` (`ParquetExec`'s schema) is `(foo, _id)`, which has one than field than their `input_dfschema` (`TableScan`'s `dfschema`), our assumption is correct here, but when converting `Projection` to a `ProjectionExec`, their `input_schema` will be same as their `input_dfschema`, both are `(foo, COUNT(*))`, our assumption does not work and thus we get a panic.

With this PR, when we get different schemas, we check if it is `made_by_us`, and ONLY return an error when it is not our fault.

### CODE REVIEW

File: datafusion/physical-expr/src/planner.rs [hunk 1]
```diff
@@ -51,20 +51,23 @@ pub fn create_physical_expr(
...
-    if input_dfschema.field_with_name(None, "_id").is_ok() {
-        if input_schema.fields.len() != input_dfschema.fields().len() {
+    if input_schema.fields.len() != input_dfschema.fields().len() {
+        // If `input_schema` and `input_dfschema` get different schemas, then
+        // there is one case where such a inconsistency is allowed, i.e., the
+        // `input_schema` has an extra `_id` field, which is added by us.
+
+        // Is this inconsistency made by us, if not, panic
+        let made_by_us = input_schema.field_with_name("_id").is_ok()
+            && input_dfschema.field_with_name(None, "_id").is_err();
```
Comment (medcl) [inline datafusion/physical-expr/src/planner.rs:61]: mostly LGTM, but I don't like we hard coded `_id` here, if we mount existing parquet files, there is a chance to have conflicts with existing schema.

Comment (SteveLauC) [reply datafusion/physical-expr/src/planner.rs:61]: #2 has already enforced that the underlying Parquet files must have an `_id` field, should we revert that as well? i.e., when tweaking the generated physical plan, only add an `_id` field if it is actually present in the Parquet file, if they don't have any `_id` field, the documents are not treated as deleted

Something to note, given that this is a hack to DataFusion, it is not that easy to make it a general implementation (having an interface where one can specify the field used for deletion at runtime)

Comment (medcl) [reply datafusion/physical-expr/src/planner.rs:61]: Just make the `_id` field configurable, allow to specify which field will be treated as the key field when we mount a general Parquet datasource, in our's case, it is `_id` field, I don't think there is a issue for our hack, we did't change the data strucute under the hood.

Comment (medcl) [reply datafusion/physical-expr/src/planner.rs:61]: I am OK with merge this PR first, to keep it more flexiable can be another PR, it is low priority, before we support mount external datasource.

Comment (SteveLauC) [reply datafusion/physical-expr/src/planner.rs:61]: > Just make the _id field configurable, allow to specify which field will be treated as the key field when we mount a general Parquet datasource

I agree this is better

>  I don't think there is a issue for our hack, we did't change the data strucute under the hood.

Maybe we can store it in the `SessionContext`, and make it configurable through a public interface

> I am OK with merge this PR first, to keep it more flexiable can be another PR, it is low priority, before we support mount external datasource.

Ok, let's merge this PR as-is then

**Review by loichyan — APPROVED** [review]

**Pull request merged**